mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp

```diff
@@ -365,27 +383,37 @@ void CLeaderboardsTimes::LoadLocalTimes(KeyValues* kv)
             char pReplayPath[MAX_PATH];
             V_ComposeFileName(RECORDING_PATH, pFoundFile, pReplayPath, MAX_PATH);
 
-            CMomReplayBase *pBase = g_ReplayFactory.LoadReplayFile(pReplayPath, false);
-            Assert(pBase != nullptr);
-
-            if (pBase)
-                m_vLocalTimes.InsertNoSort(pBase);
+            FileAsyncRequest_t request;
+            request.pszFilename = pReplayPath;
+            request.pContext = this;
+            request.pfnCallback = LocalTimesCallbackForwarder;
+            request.pszPathID = "MOD";
+            filesystem->AsyncRead(request);
 
             pFoundFile = filesystem->FindNext(found);
         }
 
         filesystem->FindClose(found);
 
-        if (!m_vLocalTimes.IsEmpty())
-        {
-            m_vLocalTimes.RedoSort();
-            m_bTimesNeedUpdate[TIMES_LOCAL] = false;
-        }
+        m_bTimesNeedUpdate[TIMES_LOCAL] = false;
+        m_flTimesLastUpdate[TIMES_LOCAL] = gpGlobals->curtime;
     }
+}
+
+void CLeaderboardsTimes::LocalTimesCallback(const FileAsyncRequest_t &request, int nBytesRead, FSAsyncStatus_t err)
+{
+    if(!request.pData || err != FSASYNC_OK)
+        return;
+
+    CUtlBuffer reader;
+    reader.SetExternalBuffer(request.pData, nBytesRead, nBytesRead, CUtlBuffer::READ_ONLY);
 
-    // Convert
-    if (!m_vLocalTimes.IsEmpty())
-        ConvertLocalTimes(kv);
+    auto replay = g_ReplayFactory.LoadReplayFromBuffer(reader, false);
+    if(replay)
+    {
+        m_vLocalTimes.Insert(replay);
```

Comment:
I worry about race conditions surrounding this.

What happens if the user hits tab to load these and leaves the map?

Is there any time where this vector would be cleared while being filled asynchronously? Since there's no mutex, is this thread safe?

Reply:
Yeah, I guess so. I think I will add another vector that stores all pending async request handles and aborts them in the panel's destructor. That should solve the first one, right?

For the second one a queue that gets filled by the callback and emptied when filling the scoreboard should do the trick?

Reply:
And while emptying the queue I should probably also suspend all async io

Reply:
Yeah I figured AsyncRead returns some sort of identifier to do the cancelling, but the panel isn't destroyed, the level is shutdown. There should be a hook for it, if there's not, we may want to pull async local map replay loading into its own system so that the map selector can also hook into it for its parsing it could do.

Reply:
Pulling it out seems like a good option, this would belong in the Replay Factory then, right?

Reply:
Yeah, as that's shared code so it'd be able to be used in the server as well if need be, but can for now exist under a `#ifdef CLIENT_DLL` block inside of it since only panels are showing loaded times.